packages/library/src/compute-limit-internal.ts

```diff
@@ -0,0 +1,188 @@
+import { Address } from '@solana/addresses';
+import { getU32Encoder } from '@solana/codecs';
+import { SOLANA_ERROR__TRANSACTION__FAILED_TO_ESTIMATE_COMPUTE_LIMIT, SolanaError } from '@solana/errors';
+import {
+    IInstruction,
+    IInstructionWithData,
+    isInstructionForProgram,
+    isInstructionWithData,
+} from '@solana/instructions';
+import { Rpc, SimulateTransactionApi } from '@solana/rpc';
+import { Blockhash, Commitment, Slot } from '@solana/rpc-types';
+import {
+    appendTransactionMessageInstruction,
+    CompilableTransactionMessage,
+    isDurableNonceTransaction,
+    isTransactionMessageWithBlockhashLifetime,
+    ITransactionMessageWithFeePayer,
+    setTransactionMessageLifetimeUsingBlockhash,
+    TransactionMessage,
+} from '@solana/transaction-messages';
+import { compileTransaction, getBase64EncodedWireTransaction } from '@solana/transactions';
+
+type ComputeUnitEstimateForTransactionMessageConfig = Readonly<{
+    abortSignal?: AbortSignal;
+    commitment?: Commitment;
+    minContextSlot?: Slot;
+    rpc: Rpc<SimulateTransactionApi>;
+    transactionMessage: CompilableTransactionMessage | (ITransactionMessageWithFeePayer & TransactionMessage);
+}>;
+
+const COMPUTE_BUDGET_PROGRAM_ADDRESS =
+    'ComputeBudget111111111111111111111111111111' as Address<'ComputeBudget111111111111111111111111111111'>;
+// HACK: Since the `compileTransaction()` method will not compile a transaction with no lifetime we
+// supply a dummy lifetime.
+const INVALID_BUT_SUFFICIENT_FOR_COMPILATION_BLOCKHASH = {
+    blockhash: '11111111111111111111111111111111' as Blockhash,
+    lastValidBlockHeight: 0n, // This is not included in compiled transactions; it can be anything.
+} as const;
+const SET_COMPUTE_UNIT_LIMIT_INSTRUCTION_INDEX = 0x02;
+
+function createComputeUnitLimitInstruction(units: number): IInstruction<typeof COMPUTE_BUDGET_PROGRAM_ADDRESS> {
+    const data = new Uint8Array(5);
+    data[0] = SET_COMPUTE_UNIT_LIMIT_INSTRUCTION_INDEX;
+    getU32Encoder().write(units, data, 1 /* offset */);
+    return Object.freeze({
+        data,
+        programAddress: COMPUTE_BUDGET_PROGRAM_ADDRESS,
+    });
+}
```

Comment:
Unless it's exposed somewhere else, it might be nice to expose this too, and add an example of adding the compute unit limit instruction after getting the units consumed.

Reply:
It's actually here:

https://github.com/solana-program/compute-budget/blob/f425ee62857b3154c85dc0f551898d5cfef67fa8/clients/js/src/generated/instructions/setComputeUnitLimit.ts#L78-L95

I would have used that actual function here, but unfortunately it would cause a circular dependency.

Reply:
Gotcha, that works